### DISCUSSION TITLE
Custom timers with Berry - How to load at startup?

### DISCUSSION THREAD

**Discussion by SimonFili** [arendst/Tasmota#17010] [Q&A] [answered]
Hi all,

I'm trying to set up a Tasmota device with complex timers.  For example, turn on a relay every 5 mins for 30 seconds.

I can do that easy with Home assistant, but I have some use case where I need to do this, with a single Tasmota device, nothing else.

So, I looked at the integrated timers of Tasmota, but not good for my use case.

I have looked at Berry, I think I can do it there, with functions and Cron.

But the issue I have is those functions and Cron are gone after a reboot.

I have tried saving them to the file system, but did not find a way to "autoload" them at boot.

Is there a solution to this?

Thanks!  Simon

**Comment by sfromis** [comment, marked as answer]
To load things at boot, you can load your code from `autoexec.be`, or even put the code directly in that file.

**Reply by SimonFili** [reply]
Thanks, now I can find this https://tasmota.github.io/docs/UFS/#autoexecbe  Was expecting to find it in the Berry sections.  Super!

**Reply by ozett** [reply]
this information how to autostart berry scripts seems to be missing in the docu section of berry.

i found it here in the discussion and not in the berry sections. 
i exptected it within the berry docu, not so [far away in the filesystem docu](https://tasmota.github.io/docs/UFS/#autoexecbe).
 thats a most valuable information and maybe should be included in the berry documentation ?

**Reply by barbudor** [reply]
Please add it
You can suggest changes by clicking on the pencil button on top right

**Reply by ozett** [reply]
never did such a thing before. but added two sentences. 
could not find a button to add a link.

i am afraid, i dont find the usual tone .. but it think you can look it up here:
https://github.com/tasmota/docs/compare/master...ozett:docs:patch-1

**Reply by barbudor** [reply]
Cool, now you need to click on the button "Create a pull-request", Give a tittle to your pull-request, save
And we will review it, and merge
Thanks

**Reply by ozett** [reply]
woohhh, thrilling. i did that first PR.
![image](https://user-images.githubusercontent.com/5937811/218024882-c99fe66a-5d4c-4ad7-8573-20784c1a4033.png)

**Reply by barbudor** [reply]
And... Merged !
Thanks for your (first) contribution

**Comment by barbudor** [comment]
> turn on a relay every 5 mins for 30 seconds.
Simple to do with Pulsetime and a rule
```
Pulsetime 130
Rule1 on timeinute|5 do power on end on
Rule1 1
```

**Reply by SimonFili** [reply]
barbudor Wow, I never used those commands on Tasmota, only used MQTT and controlled everything with Home Assistant.  Quite powerful stuff there!

1) **pulsetime1 130**
 - This works fine, I can turn on the relay manually through the GUI or with "power1 on" and 30 seconds later, the relay goes OFF

2) **rule1 on Time#Minute do power1 on endon**
 Relay1 never goes ON, I read and checked this doc: https://tasmota.github.io/docs/Rules/ and cannot find what's wrong with it.

Any help / pointers would be appreciated.

**Reply by SimonFili** [reply]
Oh also, if I can make this works, means I can do what I need with an old SonOff ESP8266 instead of an ESP32 device!

**Reply by barbudor** [reply]
Something went wrong in the formatting of my message
Should be `time#minute|5` to trigger every 5 minutes (the vertical bar is the modulo operator)

**Reply by SimonFili** [reply]
I tried time#minute|5, does not work.  So I switch to time#minute to get a faster test.  
09:58:26.709 CMD: rule1 on time#minute do power on endon

This does not work.

I can do on the console "power on" "power1 on" or off... works fine, the relay switchs

I tried many variations.  Seems like the trigger does not work.  Is there something to enable rules?

**Reply by sfromis** [reply]
Yes, barbudor already gave you the command to enable a rule set: `Rule1 1`

**Reply by SimonFili** [reply]
@sfromis Ah, this was to enable!  Thanks! ;-) 

WORKS like a charm!

Here's my notes to get this working:

```
#
##  GOAL: Have a relay turn ON every 5 minutes for 30 seconds
#
# Use pulsetime command to set the number of second to keep a relay ON
# https://tasmota.github.io/docs/Commands/#control
# for 30 seconds, you need to add 100

pulsetime1 130 

# Create rule #1 ON a 5 minutes interval, DO power1 on 
# rules https://tasmota.github.io/docs/Rules/#rule-syntax
# commands : https://tasmota.github.io/docs/Commands/#how-to-use-commands

rule1 on Time#Minute|5 do power1 on endon

# Enable rule
# After defining a rule set, you have to enable it (turn it on) using Rule<x> 1. Similarly you can disable the rule set using Rule<x> 0.
rule1 1
```

**Reply by SimonFili** [reply]
Works fine also after a restart, compared to Berry, where you need a load an autoexec file and only works on ESP32.

Thanks again to all for the help.

**Reply by sfromis** [reply]
Yeah, for small simple things, rules can be great and easy to set up. Where Berry shines is when things become a little more complex, as it is a "real programming language" with lots of features available.
